### ISSUE TITLE
Lint to warn when possibly long lived closures capture unexpected variables on the VM

### ISSUE THREAD

**Issue by jacob314** [dart-lang/linter#2569]
**Describe the rule you'd like to see implemented**
Having closures capture Dart objects is a common source of memory leaks in Flutter applications. Harmless looking code can lead to memory leaks because due to a long lived closure that capture variables that Dart users do not expect them to capture to to implementation details of the DartVM.

Dart2JS does not suffer from the same memory leaks and it is possible that the DartVM will be optimized in the future to avoid this sort of leak at which point this lint can be removed.

Common cases:
PageRouteBuilder widgets are a good first case to lint for as the closure defining the builder is virtually guaranteed to live on past when the widget that created the route builder should.

Related case: `BuildContext` objects should not be kept alive past an async gap (see https://github.com/dart-lang/linter/issues/2206) or otherwise used after the buildContext might have been unmounted.

In the general case it is challenging to distinguish between variables captured by a closure that will not lead to memory leaks and ones that will.

```dart

**Examples**
```dart
class MemoryTest extends StatelessWidget {
  final MyBigObject bigObj = MyBigObject();

  @override
  Widget build(BuildContext context) {
     var foo = 42;
      return OutlinedButton(
      child: Text("PageRouteBuilder"),
      onPressed: () {
        Navigator.of(context).pushReplacement(PageRouteBuilder(
        pageBuilder: (BuildContext context, Animation<double> animation, Animation<double> secondaryAnimation) => MemoryTest(),
        ));
       onCaptureThis: () {
          print("This=$this, foo=$foo");
      },
    );
  }
}
```
In this example, the pageBuilder closure actually captures both `context` and `this` (MemoryTest). That is surprising to the point that advanced users are unlikely to be able to root cause the issue. The problem is that the VM implementation will capture those variables as part of the pageBuild closure because of the unrelated `onCaptureThis` closure that happens to capture `this` and that the `onPressed` closure captures `context`.

The onCaptureThis closure wouldn't generate the lint as we don't have evidence that the callback from the onCaptureThis closure is long lived.

**Additional context**

**Comment by goderbauer** [comment]
I am wondering what the expected call to action for this lint would be? How do we explain to our users how to restructure the code to avoid the problem (I am not even sure the code can be restructured in all cases to avoid it...)?

**Comment by bwilkerson** [comment]
If it can't be restructured to avoid the problem, then I would consider that to be a false positive. If we can't avoid almost all false positives, then we shouldn't have the lint.

**Comment by jacob314** [comment]
I believe the code can be restructured in almost all cases to avoid the error by adding an extra static method to control exactly which members are bound. This is assuming I am understanding the error correctly.  @mraleph please correct me if I am wrong. It would be extremely useful to have a spec detailing exactly what the VM's current behavior is.

**Comment by mraleph** [comment]
Yeah, moving code into a separate method to break context chain would usually solve the problem.
